DifferentiationInterface/ext/DifferentiationInterfaceMooncakeExt/differentiate_with.jl

```diff
@@ -0,0 +1,22 @@
+@is_primitive MinimalCtx Tuple{CoDual{<:DI.DifferentiateWith},CoDual{<:AbstractArray}}
+@is_primitive MinimalCtx Tuple{CoDual{<:DI.DifferentiateWith},CoDual{<:Number}}
+
+function Mooncake.rrule!!(dw::CoDual{<:DI.DifferentiateWith}, args::CoDual...)
+    primal_func = Mooncake.primal(dw)
+    primal_args = map(arg -> Mooncake.primal(arg), args)
+
+    (; f, backend) = primal_func
+    y = f(primal_args...)
+
+    prep_same = DI.prepare_pullback_same_point_nokwarg(
+        Val(true), f, backend, primal_args..., (y,)
+    )
+
+    function pullback!!(dy)
+        tx = DI.pullback(f, prep_same, backend, primal_args, (dy,))
+        args_rdata = map((x) -> (x, Mooncake.zero_rdata(x)), only(tx))
+        return NoRData(), args_rdata...
```

Comment:
I don't think this is correct:
- if `x isa AbstractArray` you need to increment its `FData` with `dx = only(tx)`, and it has `NoRData()`
- if `x isa Number` it has `NoFData()` and you need to return `dx` as its `RData`